### PR TITLE
Use NFI for Dir and pass almost all specs

### PR DESCRIPTION
**Pull request by eregon**

### CODE REVIEW

File: src/main/ruby/core/dir.rb [hunk 1]
```diff
@@ -49,22 +53,53 @@ def initialize(path, options=undefined)
       enc = Rubinius::Type.coerce_to_encoding enc if enc
     end
 
-    Truffle.invoke_primitive :dir_open, self, path, enc
+    @encoding = enc || Encoding.find('filesystem')
```
Comment (nirvdrum) [inline src/main/ruby/core/dir.rb]: Did we ever end up adding our own `Encoding::FILESYSTEM`?

File: src/main/ruby/core/dir.rb [hunk 2]
```diff
@@ -49,22 +53,53 @@ def initialize(path, options=undefined)
...
+
+    @ptr = Truffle::POSIX.opendir(@path)
+    Errno.handle if @ptr.null?
+  end
+
+  private def ensure_open
```
Comment (nirvdrum) [inline src/main/ruby/core/dir.rb]: I think this would look a bit better as `ensure_open!`

File: src/main/ruby/core/dir.rb (hunk 1 above)
Comment (eregon) [reply src/main/ruby/core/dir.rb]: No, but I was hesitating to do it in this PR, should I?

File: src/main/ruby/core/dir.rb (hunk 2 above)
Comment (eregon) [reply src/main/ruby/core/dir.rb]: It's consistent with `IO#ensure_open` and I think `!` should only be used if there is a `!`-less variant in MRI naming conventions.
But of course this is our internal API, so I can change if you want.

File: src/main/ruby/core/dir.rb (hunk 1 above)
Comment (nirvdrum) [reply src/main/ruby/core/dir.rb]: Saving it for later is fine by me.

File: src/main/ruby/core/dir.rb (hunk 2 above)
Comment (nirvdrum) [reply src/main/ruby/core/dir.rb]: The other convention I've seen is to do it if there's a side effect or exception raised. I think that's mostly done in Rails. I don't feel strongly enough about it though.

File: src/main/ruby/core/dir.rb (hunk 1 above)
Comment (eregon) [reply src/main/ruby/core/dir.rb]: Done.

**Pull request merged**